stl/inc/xlocnum

```diff
@@ -251,6 +251,230 @@ __PURE_APPDOMAIN_GLOBAL locale::id numpunct<_Elem>::id;
+int _Parse_fp_with_ios_base(char* const _Ac, _InIt& _First, _InIt& _Last, ios_base& _Iosbase, int* _Phexexp) {
+    // get floating-point field from [_First, _Last) into _Ac
+    const auto& _Punct_fac = _STD use_facet<numpunct<_Elem>>(_Iosbase.getloc());
+    char* _Ptr             = _Ac;
+
+    constexpr size_t _Offset_dec_digit_end = 10;
+    constexpr size_t _Offset_hex_digit_end = 22;
+    constexpr size_t _Offset_neg_sign      = 22;
+    constexpr size_t _Offset_pos_sign      = _Offset_neg_sign + 1;
+    constexpr size_t _Offset_upper_x       = 24;
+    constexpr size_t _Offset_lower_x       = _Offset_upper_x + 1;
+    constexpr size_t _Offset_upper_p       = 26;
+    constexpr size_t _Offset_lower_p       = _Offset_upper_p + 1;
+    constexpr size_t _Offset_upper_e       = 0xE;
+    constexpr size_t _Offset_lower_e       = 0xE + 0x10 - 10;
+    static constexpr char _Src[]           = "0123456789ABCDEFabcdef-+XxPp";
+    _Elem _Atoms[sizeof(_Src)];
+    const ctype<_Elem>& _Ctype_fac = _STD use_facet<ctype<_Elem>>(_Iosbase.getloc());
+    _Ctype_fac.widen(_STD begin(_Src), _STD end(_Src), _Atoms);
+
+    const _Elem _Positive_sign = _Atoms[_Offset_pos_sign];
+    const _Elem _Negative_sign = _Atoms[_Offset_neg_sign];
+    const _Elem _Zero_wc       = _Atoms[0];
+
+    if (_First != _Last) {
+        if (*_First == _Positive_sign) { // gather plus sign
+            *_Ptr++ = '+';
+            ++_First;
+        } else if (*_First == _Negative_sign) { // gather minus sign
+            *_Ptr++ = '-';
+            ++_First;
+        }
+    }
+
+    *_Ptr++ = '0'; // backstop carries from sticky bit
+
+    bool _Parse_hex                = false;
+    char _Initial_dec_leading_zero = '\0';
+    if (_First != _Last && *_First == _Zero_wc) {
+        ++_First;
+        if (_First == _Last) { // "0" only
+            *_Ptr = '\0';
+            return 0;
+        }
+        if (*_First == _Atoms[_Offset_lower_x] || *_First == _Atoms[_Offset_upper_x]) { // 0X or 0x
+            _Parse_hex = true;
+            ++_First; // discard 0X or 0x for further parsing
+            *_Ptr++ = 'x';
+        } else {
+            ++_Initial_dec_leading_zero;
+        }
+    }
+
+    const int _Max_sig_dig = (*_Phexexp == _ENABLE_V2_BEHAVIOR ? _MAX_SIG_DIG_V2 : _MAX_SIG_DIG_V1);
+
+    size_t _Idx;
+    bool _Sticky           = false;
+    bool _Seendigit        = false; // seen a digit in input
+    bool _Bad              = false;
+    int _Significant       = 0; // number of significant digits
+    int _Power_of_rep_base = 0; // power of 10 or 16
+
+    const size_t _Offset_digit_end = _Parse_hex ? _Offset_hex_digit_end : _Offset_dec_digit_end;
+    const string _Grouping         = _Punct_fac.grouping();
+    if (_Grouping.empty()) {
+        for (; _First != _Last && (_Idx = _STD _Find_elem(_Atoms, *_First)) < _Offset_digit_end;
+             _Seendigit = true, (void) ++_First) {
+            if (_Max_sig_dig <= _Significant) {

```

Comment:
Style: This appears as `_Significant >= _Max_sig_dig` below, which I believe is clearer. (In modern code, we usually avoid "Yoda conditions" where the varying value appears on the right, unless there is a strong reason to phrase them that way, e.g. needing to use less-than.)